File: backend/app/application/services/vehicle_service.py
from typing import List, Optional

from app.core.uow import AbstractUnitOfWork
from app.domain.models.charger import ChargerEntity
from app.domain.models.vehicle import VehicleEntity
from app.domain.rules.compatibility_rules import find_compatible_chargers
from app.domain.rules.vehicle_rules import validate_vehicle


class VehicleService:
    def __init__(self, uow: AbstractUnitOfWork):
        self.uow = uow

    def create_vehicle(self, vehicle: VehicleEntity, current_user_id: int) -> VehicleEntity:
        with self.uow:
            user = self.uow.users.get(current_user_id)
            if not user:
                raise LookupError("User not found")

            vehicle.user_id = current_user_id
            validate_vehicle(vehicle)
            existing_vehicle = self.uow.vehicles.get_all_vehicles_by_plate(vehicle.plate)
            if existing_vehicle:
                raise ValueError("Vehicle with this plate already exists")
            new_vehicle = self.uow.vehicles.add(vehicle)
            self.uow.commit()
            return new_vehicle

    def get_all_vehicles(self) -> List[VehicleEntity]:
        with self.uow:
            return self.uow.vehicles.list()

    def get_user_vehicles(self, user_id: int) -> List[VehicleEntity]:
        with self.uow:
            user = self.uow.users.get(user_id)
            if not user:
                raise LookupError("User not found")
            return self.uow.vehicles.list_by_user_id(user_id)
    def get_vehicle(self, vehicle_id: int) -> Optional[VehicleEntity]:
        with self.uow:
            return self.uow.vehicles.get(vehicle_id)

    def get_compatible_chargers(self, vehicle_id: int) -> Optional[List[ChargerEntity]]:
        with self.uow:
            vehicle = self.uow.vehicles.get(vehicle_id)
            if not vehicle:
                return None
            chargers = self.uow.chargers.list()
            return find_compatible_chargers(vehicle, chargers)

    def delete_vehicle(self, vehicle_id: int) -> bool:
        with self.uow:
            vehicle = self.uow.vehicles.get(vehicle_id)
            if not vehicle:
                return False
            self.uow.vehicles.delete(vehicle)
            self.uow.commit()
            return True

    def update_vehicle(self, vehicle: VehicleEntity) -> VehicleEntity:
        with self.uow:
            validate_vehicle(vehicle)
            self.uow.vehicles.update(vehicle)
            self.uow.commit()
            return vehicle

File: frontend/lib/data/services/vehicle_service.dart
import '../api/api_client.dart';
import '../models/charger.dart';
import '../models/vehicle.dart';
import 'json_helpers.dart';

class VehicleService {
  VehicleService({ApiClient? apiClient})
    : _apiClient = apiClient ?? ApiClient();

  final ApiClient _apiClient;

  Future<Vehicle> createVehicle(VehicleInput input) async {
    return Vehicle.fromJson(
      parseObject(await _apiClient.post('/vehicles/', body: input.toJson())),
    );
  }

  Future<List<Vehicle>> getVehicles() async {
    return parseList(await _apiClient.get('/vehicles/'), Vehicle.fromJson);
  }

  Future<Vehicle> getVehicle(int vehicleId) async {
    return Vehicle.fromJson(
      parseObject(await _apiClient.get('/vehicles/$vehicleId')),
    );
  }

  Future<List<Charger>> getCompatibleChargers(int vehicleId) async {
    return parseList(
      await _apiClient.get('/vehicles/$vehicleId/compatible-chargers'),
      Charger.fromJson,
    );
  }

  Future<Vehicle> updateVehicle(int vehicleId, VehicleInput input) async {
    return Vehicle.fromJson(
      parseObject(
        await _apiClient.put('/vehicles/$vehicleId', body: input.toJson()),
      ),
    );
  }

  Future<void> deleteVehicle(int vehicleId) {
    return _apiClient.delete('/vehicles/$vehicleId');
  }
}
